charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py

```diff
@@ -82,44 +130,109 @@ def __init__(self, message):
 
 
 class GatewayRequirer(Object):
-    def __init__(self, charm, relation_name: str = RELATION_NAME):
-        super().__init__(charm, relation_name)
-        self.charm = charm
+    """Base class that represents a requirer relation end.
+
+    Args:
+        requirer_charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        requirer_charm (CharmBase): variable for storing the requirer application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, requirer_charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(requirer_charm, relation_name)
+        # TODO: remove this attribute, we are not really using it at all.
+        # Leaving it to keep backwards compatibility.
+        self.requirer_charm = requirer_charm
         self.relation_name = relation_name
 
-    def get_relation_data(self):
-        if not self.model.unit.is_leader():
-            return
-        gateway = self.model.relations[self.relation_name]
-        if len(gateway) == 0:
+    @staticmethod
+    def _relation_preflight_checks(relation: Relation) -> None:
+        """Series of checks for the relation and relation data.
+        Args:
+            relation (Relation): the relation object to run the checks on
+        Raises:
+            GatewayRelationDataMissingError: if data is missing or incomplete
+            GatewayRelationMissingError: if there is no related application
```

Comment:
```suggestion
            GatewayRelationDataMissingError: if data is missing or incomplete
            GatewayRelationMissingError: if there is no related application
            ops.model.TooManyRelatedAppsError: if there is more than one related application
```

Reply:
@ca-scribner good question if we should tell about raises from other lib.

But from other side, since we applied the method and allow to pass the error through I am +1

Reply:
yeah I think where it is enforcing a key function (that this requirer shouldn't have more than 1 app) we should probably say so here